Chapter 3 - Secure Communication (#23)
shresthagrawal/jslibp2p-chat-tutorial: 3/introduction.md

> ##Secure Communication

Reviewer: Check https://jslibp2p-chat-tutorial-git-securecomms.shresthagrawal.vercel.app/#/3/introduction, needs a space.
```suggestion
## Secure Communication
```

> ## Secure Communication

Reviewer 2: ```suggestion
Secure Connection
==============
```

> As has been noted several times throughout this tutorial, libp2p is a modular network stack and one of the module types available for libp2p is the [connection encryption](https://github.com/libp2p/js-libp2p/blob/master/doc/CONFIGURATION.md#connection-encryption) module type, which is set using the `connEncryption` option when configuring a node.  As the name suggests, this module deals with how data that is sent between two nodes is encrypted using asymmetric encryption.  

Reviewer 2: ```suggestion
Libp2p has [connection encryption](https://github.com/libp2p/js-libp2p/blob/master/doc/CONFIGURATION.md#connection-encryption) interface to easily plug use any encryption method.  This can be configured by setting the `connEncryption` in node options.  This will make sure every transport message is encrypted using the provide method.   
```

> Several of the transport modules that libp2p supports do not have native encryption so data would be sent in plain text between nodes.  At the moment, our chat application is configured this way and sends data as plain text which is exactly what it sounds like.  When one node sends data to another, it just sends the messages as plain text, so anyone can intercept or read the data being sent over whatever transport module is being used (i.e. TCP/UDP/Websockets/WebRTC) can read everything that's sent.  I don't know about you but I don't think a chat app is very useful if everything I send to someone else can be read by anyone and then posted all over the internet.  

Reviewer 2: ```suggestion
Several of the transport modules that libp2p supports do not have native encryption so data would be sent in plain text between nodes.  At the moment, our chat application is configured this way and sends data as plain text which is exactly what it sounds like. I don't think anyone would like to use a chat system where everyone can read everyone's message. 
```

> If you want a better understanding of how encryption works and the difference between symmetric and asymmetric encryption, [here's](https://wordtothewise.com/2014/09/cryptography-alice-bob/) a good article that gives an overview through the possibly familiar story of Alice and Bob.
> 
> This is where encryption comes in.  Libp2p offers two encryption modules, [secio](https://github.com/libp2p/js-libp2p-secio) and [noise](https://github.com/NodeFactoryIo/js-libp2p-noise) though you could also conceivably build your own.  Regardless of which one you choose, under the hood, libp2p will use that encryption library to secure the communication channel between your node and the one you're talking to so that prying eyes can't see what you're saying.  Here's how it works:
>  * Your node writes data into the outbound connection stream.
>  * Libp2p encrypts that data using the public key of the node you're talking to.
>  * That encrypted message gets sent over the internet using the transport module your nodes are talking on
>  * The remote node receives the data and decrypts it using its private key
>  * The node processes the decrypted data and outputs according to whichever protocol you're using.  In the case of our chat application, it prints the message to the screen
> 
> All this happens behind the scenes so you as a user of libp2p don't have to think about it at all.  When your node initially connects to another node, it asks the other node if it supports the encryption method specified in your node's configuration.  If not, no connection will ever be established.
> 
> Now that we have a basic picture of why encryption is important and how it's used in libp2p, let's implement it in our chat application.

Reviewer 2: ```suggestion
The general flow for all libp2p encryption module goes as follows:
 * Your node writes data into the outbound connection stream
 * Libp2p encrypts that data using the public key of the node you're talking to.
 * That encrypted message gets sent over the internet using the specified transport module
 * The remote node receives the data and decrypts it using its private key
 * The remote node processes the decrypted data and outputs according to whichever protocol you're using.  In the case of our chat application, it prints the message to the screen

If you want a better understanding of how encryption works and the difference between symmetric and asymmetric encryption, [here's](https://wordtothewise.com/2014/09/cryptography-alice-bob/) a good article that gives an overview through the possibly familiar story of Alice and Bob.

Libp2p offers two encryption modules, [secio](https://github.com/libp2p/js-libp2p-secio) and [noise](https://github.com/NodeFactoryIo/js-libp2p-noise) though you could also conceivably build your own.  Regardless of which one you choose, under the hood, libp2p will use that encryption library to secure the communication channel. It is important for all the communicating nodes to use the same encryption method to establish a successful connection. 

Now that we have a basic picture of why encryption is important and how it's used in libp2p, let's implement it in our chat application.
```
